### DISCUSSION TITLE
Is there anyway to generate the 3d array of density

### DISCUSSION THREAD

**Discussion by applejwjcat** [osyris-project/osyris#114] [General]
If I use dataset.load(), I can get 1d array of densitywith the shape of size**3 use 
```
density = data['hydro']['density'].values
```

but I could't transform it into 3d array directly use np.reshape(size,size,size), how can I obtain the 3d array, thanks !

**Comment by nvaytet** [comment]
I tried:
```Py
import numpy as np
import osyris as oo

a = np.random.random((10, 15, 20))
ar = oo.Array(a.flatten(), unit='m')
ar
```
gives
```
'' Min: 1.335e-04 Max: 1.000 [m] (3000,)
```

and then I can reshape using either
```Py
np.reshape(ar, (10, 15, 20))
```
or
```Py
ar.reshape(10, 15, 20)
```

Both give
```
'' Min: 1.335e-04 Max: 1.000 [m] (10, 15, 20)
```

So you should be able to reshape the osyris arrays.
Can you give a bit more information on what you Array is? What shape it has and what are you trying to reshape it to?

**Reply by applejwjcat** [reply]
I was sorry that I didn't illustrate my problem clearly, I have ran a 256^3 ( levelmin = 6, levelmax = 8) galactic disk simulation, I use your code to read all the output and get the ```data``` variable, I also use the ramses default analysis code like ```amr2cube``` to obtain the 3d variable array(like ```density```). I call it ```cube_dens``` now.

As I mentioned above, I could't reshape the ```data['amr']['density'].values``` use just ```reshape``` method to obtain the same 3d array containing a full cube density field. It was not equal to ```cube_dens``` variable, the element in them have different order. 

for example, the ```cbue_density``` like that(I use the flatten method) : 
```
print(cube_dens.flatten()[:10])
array([3.4921446e-26, 3.4851280e-26, 3.4842430e-26, 3.4843792e-26,
       3.4872459e-26, 3.4936394e-26, 3.5037729e-26, 3.5180223e-26,
       3.5365503e-26, 3.5593863e-26], dtype=float32)
```
but the ```osyris_dens``` like this:
```
print(osyris_dens[:10])
array([3.49214474e-26, 4.00026503e-26, 4.31842301e-26, 4.27960277e-26,
       4.27802077e-26, 4.52052012e-26, 5.41038900e-26, 6.29158792e-26,
       3.78942235e-26, 3.99364913e-26])
```
The first element in these two array is the same, but the other is not. When just reshape the ```osyris_dens``` , I could not obtain the full cube of density field, the element is disorder.

I need to use the full 3d variable density for post-analyse, is there anyway to do that?
Thanks for develop the wonderful codes and open to us. Any suggestion is welcome!

**Comment by nvaytet** [comment]
Hi again,

One of the main motivations for creating `osyris` is so that one did not need to use `amr2cube` or `amr2map` to make plots.
Instead, we want to take advantage of the AMR mesh to load only the data that we need.
In essence, we are loading only the leaf cells, in a (semi) random order.
Osyris then sees each cell as a unique little cube in space, the cells are not aware that they are part of a grid, they just have a position and a size. This saves a **lot** of memory, compared to using `amr2cube` which fills the entire space with a dense 3d array.

My first question is: are you sure you really need a dense 3d cube for your post-analysis?
It is always a good idea to question the requirements. If you are feeding the data into e.g. a ray tracing code that only accepts a dense 3d cube as an input, then yes, you cannot get around making the 3d cube.
On the other hand, if you need to make some other visualization or integrate some quantities, try to see if you cannot do it keeping the data unstructured/raw.

If you are in the first case, and you really need the 3d cube, I suggest two options:

1. Just use `amr2cube`, this may actually do the job that you want fine.

2. If you want to use `osyris` for this (because you want to stay in python, or you need the physical units, or some other reason), you could make a stack of slices using `map` to fill in a master cube of data. In the following example, I find the densest cell in my simulation, and then make a 128^3 cube of data around that center (of size 2000^3 AU), making 128 layers in the `z` direction.
```Py
import osyris
import numpy as np
data = osyris.Dataset(8, path="osyrisdata/starformation").load()

ind = np.argmax(data["hydro"]["density"])
center = data["amr"]["position"][ind]

dx = 2000. # in au
N = 128  # resolution
cube = np.zeros([N, N, N])  # master array
step = dx / N
lower_z = center - osyris.Vector(0, 0, 0.5*dx, unit='au')

for i in range(N):
    origin = lower_z + osyris.Vector(0, 0, (i + 0.5)*step, unit='au')

    p = osyris.map({"data": data["hydro"]["density"]},
                   dx=dx * osyris.units('au'),
                   origin=origin,
                   direction="z",
                   resolution=N,
                   plot=False)  # use plot=False to not make the figure but just return the data

    cube[..., i] = p.layers[0]['data']
```

Method 2 might be quite slow, depending on what you choose for resolution, but if you don't have a million outputs to process, it might be ok?

If it's not, if have some ideas on how we could modify `map` to speed this up.
Hope this helps!

**Reply by applejwjcat** [reply]
Thanks for your reply!!!
I have understood it ! Indeed, in my work, I use ```osyris```  to plot the slice figure mostly. But I encountered a problem when analyzing the galactic disk simulation data. That's why I want to obtain the cube data. Here is my problem, I plot the figure according to the ```osyris``` manual like that:
``` python
path = '/home/tristan/Documents/code/ramses_final/'
data = osyris.Dataset(11, path=path).load()
ind = np.argmax(data["hydro"]["density"])
center = data["amr"]["position"][ind.values]
```
I can load the data rightly
```
Processing 64 files in /home/tristan/Documents/code/ramses_final/output_00011
 10% : read    1835008 cells,        721 particles
 20% : read    3407872 cells,       1456 particles
 31% : read    5242888 cells,       2045 particles
 40% : read    6815760 cells,       2833 particles
 50% : read    8388616 cells,       3397 particles
 60% : read   10223624 cells,       4220 particles
 70% : read   11796496 cells,       4987 particles
 81% : read   13631496 cells,       5687 particles
 90% : read   15204360 cells,       6512 particles
Loaded: 16777216 cells, 7213 particles.
```
But when I plot the figure,
```python
osyris.map(dat["hydro"]["density"],
           norm="log",
           vmin = 1.087e-26, # the min value
           vmax = 9.335e-23, # the max value # if I omit the ```vmin```, ```vmax``` , got errors 
           dx= 2000 * osyris.units('pc'), # the size is right
           origin=center, direction="z")
```
It output the a figure with nothing except a colorbar. I have tested your sample data and got the right figure. I have update some of my data in the github via [ramses_data](https://github.com/applejwjcat/ramses_data), could you please help me check what's wrong with the analysis.

Otherwise, I also output the map data use flag ```plot=False```, and the data is like that :
```python
print(p.layers[0]['data'])
masked_array(
  data=[[--, --, --, ..., --, --, --],
        [--, --, --, ..., --, --, --],
        [--, --, --, ..., --, --, --],
        ...,
        [--, --, --, ..., --, --, --],
        [--, --, --, ..., --, --, --],
        [--, --, --, ..., --, --, --]],
  mask=[[ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        ...,
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True]],
  fill_value=1e+20,
  dtype=float64)
```
the ```p``` data was not loaded rightly, but  ```data['hydro']['density']``` is. So there is something wrong...

I am waiting for your reply, thanks again for your code!

**Comment by haugboel** [comment]
I have had similar problems and changing resolution or giving the slice a thickness helped. It is like for some slice points it does not find a cell to interpolate from.

Sendt fra min iPhone

**Comment by nvaytet** [comment]
It seems @haugboel might be right, giving the slice a thickness seems to help:
```Py
osyris.map({"data": data["hydro"]["density"], "norm": "log"},
           dx=1000 * osyris.units('pc'),
           dz=20 * osyris.units('pc'),
           origin=center,
           direction="z")
```
with the data you posted on github yields
![Screenshot at 2022-10-12 19-48-36](https://user-images.githubusercontent.com/39047984/195412903-8774efb7-b223-47bf-8626-7b2c952b5a10.png)

Now I need to figure out why it is not working when there is no thickness....

**Comment by nvaytet** [comment]
It turns out the issue was an integer overflow happening when large values were being passed to `numba` inside the `map` calculation.
Python can represent any integers but in `numba` we were reaching the int64 limit when passing positions in cm that have a galactic range of Mpc.

I fixed it by normalizing the values that I am passing to `numba`. See #115

**Reply by nvaytet** [reply]
This shows the importance of reporting such things, because now we have one more bug fixed ;-)

**Comment by nvaytet** [comment]
I made a new release of `osyris` with the fix: `2.10.1`. Try it out and let me know if it solved your issue.

**Reply by applejwjcat** [reply]
Fantastic work! It solved my problem. Thanks again.

**Comment by haugboel** [comment]
Fantastic; Thanks Neil.

Sendt fra min iPhone
